### PR TITLE
Throw exceptions with context

### PR DESCRIPTION
**Pull request by dmlogic**
As per discussion in #360, this introduces additional exceptions to reduce the need to examine the error message string

### CODE REVIEW

**Comment by nyamsprod** [comment]
you should run `composer test` prior to submitting your commit. 

if you had some issue with coding style you can type 

```
./vendor/bin/php-cs-fixer fix -vvv --allow-risky=yes
```

File: src/Stream.php [hunk 1]
```diff
@@ -243,7 +243,7 @@ public function appendFilter(string $filtername, int $read_write, $params = null
             return;
         }
 
-        throw new Exception(sprintf('unable to locate filter `%s`', $filtername));
+        throw new CannotAddStreamFilter(sprintf('unable to locate filter `%s`', $filtername));
```
Comment (nyamsprod) [inline src/Stream.php]: This should be a `InvalidArgument` exception. Stream Filtering is supported *but* the filter name is invalid.

File: src/AbstractCsv.php [hunk 2]
```diff
@@ -486,7 +486,7 @@ public function setOutputBOM(string $str): self
     public function addStreamFilter(string $filtername, $params = null): self
     {
         if (!$this->document instanceof Stream) {
-            throw new Exception('The stream filter API can not be used');
+            throw new CannotAddStreamFilter('The stream filter API can not be used');
```
Comment (nyamsprod) [inline src/AbstractCsv.php]: I'm wondering if we should not name this one `MissingStreamFilterSupport` or `StreamFilterSupportMissing`

Comment (dmlogic) [reply src/AbstractCsv.php]: `StreamFilterSupportMissing` sounds good to me. Want me to change that?

Comment (nyamsprod) [reply src/AbstractCsv.php]: yes please

**Pull request merged**